sklearn/utils/_pprint.py

```diff
@@ -0,0 +1,303 @@
+
+    When estimators cannot be printed on a single line, the builtin _format()
+    will call _pprint_estimator() because it was registered to do so (see
+    _dispatch[BaseEstimator.__repr__] = _pprint_estimator).
+
+    both _format_dict_items() and _pprint_estimator() use the
+    _format_params_or_dict_items() method that will format parameters and
+    key-value pairs respecting the compact parameter. This method needs another
+    subroutine _pprint_key_val_tuple() used when a parameter or a key-value
+    pair is too long to fit on a single line. This subroutine is called in
+    _format() and is registered as well in the _dispatch dict (just like
+    _pprint_estimator). We had to create the two classes KeyValTuple and
+    KeyValTupleParam for this.
+    """
+
+    def __init__(self, indent=1, width=80, depth=None, stream=None, *,
+                 compact=False, indent_at_name=True):
+        super().__init__(indent, width, depth, stream, compact=compact)
+        self._indent_at_name = indent_at_name
+        if self._indent_at_name:
+            self._indent_per_level = 1  # ignore indent param
+        self._changed_only = get_config()['print_changed_only']
+
+    def format(self, object, context, maxlevels, level):
+        return _safe_repr(object, context, maxlevels, level,
+                          changed_only=self._changed_only)
+
+    def _pprint_estimator(self, object, stream, indent, allowance, context,
+                          level):
+        stream.write(object.__class__.__name__ + '(')
+        if self._indent_at_name:
+            indent += len(object.__class__.__name__)
+
+        if self._changed_only:
+            params = _changed_params(object)
+        else:
+            params = object.get_params(deep=False)
+
+        self._format_params(params.items(), stream, indent, allowance + 1,
+                            context, level)
+        stream.write(')')
+
+    def _format_dict_items(self, items, stream, indent, allowance, context,
+                           level):
+        return self._format_params_or_dict_items(
+            items, stream, indent, allowance, context, level, is_dict=True)
+
+    def _format_params(self, items, stream, indent, allowance, context, level):
+        return self._format_params_or_dict_items(
+            items, stream, indent, allowance, context, level, is_dict=False)
+
+    def _format_params_or_dict_items(self, object, stream, indent, allowance,
+                                     context, level, is_dict):
+        """Format dict items or parameters respecting the compact=True
+        parameter. For some reason, the builtin rendering of dict items doesn't
+        respect compact=True and will use one line per key-value if all cannot
+        fit in a single line.
+        Dict items will be rendered as <'key': value> while params will be
+        rendered as <key=value>. The implementation is mostly copy/pasting from
```

Comment:
I'm not immediately convinced that we have sufficient attribution/licence here. Convince me

Reply:
Hmm good point.

The [PSF License](https://en.wikipedia.org/wiki/Python_Software_Foundation_License) is BSD-like and non-copyleft, so I assume we can use the code here?

I'm really not expert on licenses so I might be wrong.

Reply:
Most open source licences that are not copyleft merely require attribution and inclusion of the licence (what I'm not sure is whether linking to or referencing the licence is sufficient in some circumstances)